lessons/bash-intro/lesson.md

```diff
@@ -0,0 +1,330 @@
+
+```
+ls tmp
+```
+
+you should see the files in the `tmp` directory (`file_a`). Now try
+
+```
+cat tmp/file_a
+cd tmp
+cat tmp/file_a
+```
+
+What went wrong?
+
+
+## Absolute paths
+
+Sometimes, you'll need to access files from another location in the file hierarchy. With relative paths, you can use `..` to access the parent directory, so `../../src/include/` goes up two directories, then down through `src` and `include`. This can get cumbersome if you're not sure exactly how many times you need to go up. 
+
+Instead, there are absolute paths. On Unix systems, there is one root node for the file-system: `/` (compare this with `C:\` in Windows). All files are in subdirectories of root. Any pathname starting with `/` is an absolute path starting from the root.
+
+Try out `ls /`
+
+Another special location is your home directory. The short-form is `~/`. On a Mac, this would be `/Users/username/`. You can also short-form any user's home-directory with `~username/`.
+
+Try `cat ~/.bashrc`.
+
+Most of the time working with scripts, you should use relative paths, because one cannot assume the state of another's file system.
+
+
+# Globbing
+
+If you want to pass more than one filename as an argument, you can use *globbing*. The special patterns for globbing are:
+
+- `*` Match any string of arbitrary length
+- `?` Match any single character
+- `[]` Match any of the characters within square brackets.
+
+Try moving all python files into the `src` directory. How about copying every file with `a`?
+
+# Getting more powerful
+
+Some true strengths of the command line come from combining commands. This is the "Unix philosophy" of having many small programs that perform their jobs well. These units can be combined to perform more complicated tasks.
+
+# Pipes (stdin/out)
+
+Commands can be strung together with pipes `|`. Those of you using R may have seen the concept in the `dplyr` package. Commands typically have an input (stdin) and an output (stdout). The pipe attaches the output of one command into the input of another.
+
+This is easier to see with commands like `cat` and `head`. Try these out using our demo directory:
+
+```
+head animals
+cat animals
+cat animals | head
+ls | head -n2
+head
+grep 'fish' animals | wc -l
+```
+
+Can you figure out how each of these commands behave from the above?
+
+You can also redirect output into a file with `>` or `>>`. The former over-writes the file, while `>>` appends to the file if it exists. For example,
+
+```
+grep 'fish' animals > list-of-fish
+```
+
+Will create a new file with a list of fish. Files can also be used for `stdin` using the `<` character. The above command would be
+
+```
+grep fish < animals > list-of-fish
+```
+
+Or, with pipes,
+
+```
+grep fish < animals | wc -l > number-of-fish
+```
+
+Note that `<` appears after the first command, because each line must start witha  command. What would happen if you tried `animals > grep fish`?
```

Comment:
Confusingly, `<` also is a "command", `< animals grep fish` would also work (but maybe you don't need to bring that up.